### ISSUE TITLE
large webpack build almost hangs at 91% on an "additional asset processing" step

### ISSUE THREAD

**Issue by doberkofler** [webpack/webpack#4558]
**Do you want to request a *feature* or report a *bug*?**
possible bug

I have already asked on StackOverflow and it seems as if this behavior would be quite common but there have been no possible solutions.

**What is the current behavior?**
I have a large webpack build that almost hangs at 91% on an "additional asset processing" step.
To complete processing take about 8 minutes but the step "additional asset processing" consumes at lwat half of the time.
Webpack does not give me a lot more information and I would like to better understand if this is "normal", a bug or what can be done to eventually optimize my build?
```
56205ms building modules
31ms sealing
0ms optimizing
0ms basic module optimization
15ms module optimization
0ms advanced module optimization
0ms basic chunk optimization
0ms chunk optimization
16ms advanced chunk optimization
14487ms building modules
0ms module and chunk tree optimization
31ms module reviving
0ms module order optimization
16ms module id optimization
0ms chunk reviving
16ms chunk order optimization
31ms chunk id optimization
140ms hashing
0ms module assets processing
265ms chunk assets processing
0ms additional chunk assets processing
0ms recording
206740ms additional asset processing
79781ms chunk asset optimization
1ms asset optimization
906ms emitting
```

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**
faster or more information on what is currently being done

**If this is a feature request, what is motivation or use case for changing the behavior?**

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**
node: 6.10.0
webpack: 2.3.1
OS: Windows 7 x64

**Comment by sokra** [comment]
duplicate of #4550

**Comment by doberkofler** [comment]
@sokra It looks very similar but in my case, the upgrade to 2.3.1 actually improved the performance by about 10%. The performance actually started to slow down when migrating from webpack1 to 2. Is there anything I can do to track down where the time is lost to eventually help?

**Comment by igor-lemon** [comment]
At the latest v2.3.3 update wrote that 
> Fix performance issue with cheap-source-maps
But I still have the same problem.

**Comment by obykoo** [comment]
I upgrade to latest webpack due to this issue, but still same it usually takes around 10 minutes for dist build and most of this time is haning on  91% additional asset processing

**Comment by hsingh23** [comment]
I also have this problem for production builds
I use devtool: 'source-map',

**Comment by leyyinad** [comment]
Happens for me only when using `babili-webpack-plugin`

**Comment by j-nord** [comment]
It hangs on each large file at 91%. It takes a lot of time and finally provides the error message: Not enough memory.

Hardly to use.

**Comment by igor-lemon** [comment]
I found a solution. Simple add more memory for running with --max-old-space-size param
`node ... --max-old-space-size=8192`

**Comment by j-nord** [comment]
Thanks for the answer.

Thanks for the answer!

But that does not solve the problem that you have to wait on a fast computer over 10 minutes with 25% CPU usage and over 1 GB of memory each time to the completion.

That is to Long!

**Comment by dhatawesomedude** [comment]
@leyyinad I have the same issue. Mine takes about 4minutes at 91% with babili-webpack-plugin enabled.

**Comment by leyyinad** [comment]
See here: https://github.com/babel/babili/issues/332

**Comment by dhatawesomedude** [comment]
I already tried that. It fixes the initial issue of the build failing to complete, but the issue with being stuck at 91% processing still remains.

**Comment by mrampson** [comment]
Its probably UglifyJSPlugin or UglifyJsParallelPlugin

**Comment by EthanStandel** [comment]
My team's project always hangs on 91% and then sits for a bit on 92%, but removing UglifyJSPlugin fixed the 91% hang.

**Comment by matt-lough** [comment]
@EthanStandel I can also confirm that I just removed the UglifyJSPlugin and it fixed the 91% hang in my project

**Comment by doberkofler** [comment]
I originally opened this SR and can also confirm, that there have been a few minor (~10%) performance improvements in webpack but most of the time is still consumed in the UglifyJSPlugin. In my current environment webpack does not actually hang but the `additional asset processing` step can take up to 10 minutes.

**Comment by rjarmstrong** [comment]
Don't rely on Webpack plugins for everything e.g. we are using this in Go and its WAAAY faster: https://github.com/tdewolff/minify with maybe slightly less optimisations.

**Comment by crazyxhz** [comment]
I have this issue too. Finally found out that webpack 3.5.5 and Extract text plugin 3.0.0 combo will cause this problem. Each incremental dev build will take about 20 seconds which is unacceptable.
Seems like webpack is busy processing and extracting css which is not modified at all.
After I change webpack to 2.7.0 and Extract text plugin to 2.1.2, the incremental build took about 3 seconds. :stuck_out_tongue_winking_eye:

**Comment by gallowdp** [comment]
Why was this closed? I'm still having this issue after trying several different workarounds posted. I'm still seeing a lot of people commenting on various threads that are experiencing the same thing. The only way I can resolve is setting the webpack devtool to 'eval', but that's not supposed to be used for production.

**Comment by andreialecu** [comment]
I hate to +1 this, but I have been working around this problem for a long time, and I can probably provide more info:

The only way acceptable way I can get my builds to succeed is to use https://github.com/gdborton/webpack-parallel-uglify-plugin and ensure it uses UglifyJS instead of UglifyES. If I set that plugin into UglifyES mode it will hang at 91% just as the main webpack uglify plugin.

Builds **also succeed if `devtool` is set to `eval`** using either of the two plugins in `uglify-es`mode. Any other devtool doesn't work, including the `null/undefined` option, which I'm using anyway. `devtool: "eval"` builds however are huge.

I've tracked this down to `@angular/compiler`. If I remove the `require` to it in the code, builds minify properly regardless of the `devtool` setting, and with either plugin.

Now the question is why does `devtool: "eval"` make this work, but `devtool: null` doesn't?

Could this be reopened?

**Comment by lewebdalex** [comment]
I have the same issue with latest version of webpack.
Everything was running smoothly before.

**Comment by andreialecu** [comment]
I've been debugging this issue and found out it is probably a bug in UglifyJS, see: https://github.com/webpack/webpack/issues/4558#issuecomment-352170623

I'm able to avoid the 91% hang by passing `{ compress: false }` to `UglifyJSPlugin`, eg:

```javascript
new UglifyJSPlugin({
      parallel: true,
      uglifyOptions: {
        ecma: 6,
        compress: false // hangs without this
      },
      cache: path.join(__dirname, 'webpack-cache/uglify-cache'),
    })
```

**Comment by vesper8** [comment]
@andreialecu's solution worked for me, but since I'm using UglifyES first I had to install

```
"uglify-es": "^3.1.3",
"uglifyjs-webpack-plugin": "^1.0.0-beta.1",
```

And add this at the top of my webpack.mix.js file

```
const UglifyJSPlugin = require('uglifyjs-webpack-plugin');
```

I wish I could use compression though.. but it makes things 100x faster if I don't

**Comment by benjaminadrom** [comment]
is this still a thing? the solution "compress: false", works but it is accompanied by much larger files :/

**Comment by andreialecu** [comment]
This has been fixed in `uglify-es` for a few versions now as far as I can tell. This was merged in December: https://github.com/mishoo/UglifyJS2/pull/2614

Make sure your webpack, or uglifyjs-webpack-plugin is using the latest `uglify-es` version.

**Comment by guru-zoomforth** [comment]
I'm still running into this issue with the recent version of uglifyjs-webpack-plugin -- it appears to be using the most recent version of uglify-es as well. (yarn info on the package says 'uglify-es': '^3.3.4')

**Comment by guru-zoomforth** [comment]
We are still running into this problem even though our uglifyjs-webpack-plugin is using the most recent uglify-es (version 3.3.8)

**Comment by benjaminadrom** [comment]
@guru-zoomforth yes we too,
we installed "uglifyjs-webpack-plugin" -V1.1.6
and used it in the webpack.config-File like

....
const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
...
plugins: [
	...
	new UglifyJsPlugin(),
	...
]

**Comment by yantakus** [comment]
Here are results of my experiments with `"uglifyjs-webpack-plugin" -V1.1.6`:

Webpack 3.10.0 with `-p` flag, no uglifyjs-webpack plugin:
**326 000** ms build time, **150** mb bundle size

"uglifyjs-webpack-plugin" -V1.1.6 (removed the `-p` flag), no config:
**496 000** ms build time, **183** mb bundle size

uglifyjs-webpack-plugin with the following config:
```
{
      parallel: true,
      uglifyOptions: {
        ecma: 6,
      },
      cache: path.join(__dirname, 'webpack-cache/uglify-cache'),
    }
```
**315** 000 ms build time, **183** mb bundle size

Another build with same config for the cache to take effect:
**83** 000 ms build time, **183** mb bundle size

The same config, but with `compress: false,`:
**157** 000 ms build time, **185** mb bundle size.

So caching makes it much faster, `compress: false` seems to make it slower and increases the bundle. But the question is - why does pure `webpack -p` produces the smallest bundle? Am I missing some bundle size optimization options in the config?

**Comment by develra** [comment]
@yantakus - not sure how you are chunking, but I've found it really easy to duplicate a ton of dependencies if you don't start the plugin chain with a something like 
`
new webpack.optimize.CommonsChunkPlugin({
            children: true,
            async: true,
            minChunks: 3
        })
`
Adding this reduced my chunked out size from 208Mb to 6.3Mb

**Comment by aapoalas** [comment]
Having this same problem. Our devtool option is set to inline-cheap-module-source-map.

Disabling UglifyJsPlugin does the trick, build passes. Funnily enough, so does setting devtool to eval.

Even funnier, unsetting devtool does not fix the problem. What?!

**Comment by tim-soft** [comment]
I seem to be having the same issue on Webpack 4 no matter which plugin I'm using, UglifyJS, babel-minify etc, they all hang at `chunk assets processing` and sit there until I run out of RAM

**Comment by Dreamwalker420** [comment]
I experienced the same error on  ```ng serve```.  I removed the following from the ```.angular-cli.json``` file.

```
            "styles": [
                "assets/css/styles.css"

            ],
            "scripts": [
                "assets/js/project.js"      

            ],
```

**Comment by MengChen2015** [comment]
my webpack version:@2.7.0 .
I also confirm that commenting out UglifyJsPlugin(or removing) solved my problem.Before it takes out 25 seconds on " hangs at 91% on an "additional asset processing" step,now it takes out 1-3 seconds.
/* new UglifyJsPlugin({
      uglifyOptions: {
        compress: true
      },
      sourceMap: true,
    }),*/

**Comment by putninv** [comment]
Hi!

**Comment by putninv** [comment]
Hi! I am just set regExp for files extension, that I want to uglify
```
new UglifyJsPlugin({
        test: /\.js$/i,
        exclude: /node_modules/,
        uglifyOptions: {
          output: {
            comments: false
          }
        }
      })
```

This works for me. And what I found that, every next webpack build will be built much faster than previous. First my try was about 12 minutes, 6th try was about 2 minutes with the same configuration.

**Comment by gravindran2412** [comment]
Same issue:
 91% additional asset processing scripts-webpack-pluginENOENT: no such file or directory, open 'C:\Users\graveendran\eclipse-workspace\node_modules\jquery\dist\jquery.min.js'

**Comment by FaizalMalik** [comment]
I have fixed this issue by removing the **'new webpack.optimize.UglifyJsPlugin({sourceMap: true}),'**
from webpack config file - happy coding 👍

**Comment by doberkofler** [comment]
I originally opened this SR and for me the solution was to always use the latest webpack version and invoke node with the --max_old_space_size=4096 option. This permanently solved the problem for me!

**Comment by Alzatee** [comment]
En mi caso este error fue a causa de que instalé el archivo popper.js manualmente, en la dependencia del bootstrap, por lo tanto al reinstalar el node_modules, el packaje.json instaló las dependencias normalmente, excepto el popper.js, ya que ese lo metí yo a mano en la dependencia del bootstrap.

**Comment by Alzatee** [comment]
En mi caso, este error fue una causa que instalé el archivo popper.js manualmente, en la dependencia del bootstrap, por lo tanto al reinstalar el node_modules, el packaje.json instaló las dependencias normalmente, excepto el popper.js, you you that that Lo metí yo a mano en la dependencia del bootstrap.

**Comment by Alzatee** [comment]
En mi caso, este error fue una causa que instalé el archivo popper.js manualmente, en la dependencia del bootstrap, por lo tanto al reinstalar el node_modules, el packaje.json instaló las dependencias normalmente, excepto el popper.js, you you that that Lo metí yo a mano en la dependencia del bootstrap.

**Comment by Alzatee** [comment]
En mi caso, este error fue una causa que metí el archivo popper.js manualmente, en la dependencia del bootstrap, por lo tanto al reinstalar el node_modules, el packaje.json instaló las dependencias normalmente, excepto el popper.js, you you that that Lo metí yo a mano en la dependencia del bootstrap.

**Comment by Alzatee** [comment]
En mi caso, este error fue una causa que agregué el archivo popper.js manualmente, en la dependencia del bootstrap, por lo tanto al reinstalar el node_modules, el packaje.json instaló las dependencias normalmente, excepto el popper.js, ya que ese no lo instalé, sino que lo puse manualmente dentro el bootstrap

**Comment by toanbku** [comment]
Same problem when I deploy on server. In my case, I just fix it by increasing the swap ram area. 
```
sudo fallocate -l 4G /swapfile && \
sudo chmod 600 /swapfile && \
sudo mkswap /swapfile && \
sudo swapon /swapfile && \
sudo swapon -s && \
sudo cp /etc/fstab /etc/fstab.bak && \
echo '/swapfile none swap sw 0 0' | sudo tee -a /etc/fstab
```

**Comment by ramsunvtech** [comment]
Below settings helped me to reduce the heap memory since it will not include the sourceMap file.

```
devtool: false, // Webpack.

new webpack.optimize.UglifyJsPlugin({
   sourceMap: shouldUseSourceMap
}),
```

or below code will help for latest webpack uglify plugin.

```
> npm install uglifyjs-webpack-plugin --save-dev

// conf.js
const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
module.exports = {
  optimization: {
    minimizer: [
      new UglifyJsPlugin({
        sourceMap: true,
      }),
    ],
  },
};
```

If this still not help to fix the problem. below command helps to increase heap memory and answer link to know more about heap memory is below as well.

```
node --max-old-space-size=16000 scripts/build.js
```

https://stackoverflow.com/a/57480923/1533666

**Comment by Nghiemducnam** [comment]
I met the same case, you can try this statement in terminal: npm i
